From 35ee39bc1e3a73cf7c35a1dd56fd61ffaf93a6ff Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 9 Nov 2021 02:30:41 +0100
Subject: [PATCH 0323/1306] server: Continuously send pointer update messages
 while it's down.

CW-Bug-Id: #18214

server: Clear POINTER_MESSAGE_FLAG_NEW from the correct field.

CW-Bug-Id: #19994

server: Merge WM_POINTERUPDATE messages together.

CW-Bug-Id: #19994

server: Keep desktop cursor position updated while touch is set.

CW-Bug-Id: #19994
---
 server/queue.c      | 116 +++++++++++++++++++++++++++++++++++++++++++-
 server/user.h       |   2 +
 server/window.c     |   1 +
 server/winstation.c |   2 +
 4 files changed, 120 insertions(+), 1 deletion(-)

diff --git a/server/queue.c b/server/queue.c
index aa51719ee14..de276fa295a 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -655,6 +655,36 @@ static inline unsigned int get_unique_id(void)
     return id;
 }
 
+static int merge_pointer_update_message( struct thread_input *input, const struct message *msg )
+{
+    struct hardware_msg_data *prev_data, *msg_data = msg->data;
+    struct message *prev;
+    struct list *ptr;
+
+    for (ptr = list_tail( &input->msg_list ); ptr; ptr = list_prev( &input->msg_list, ptr ))
+    {
+        prev = LIST_ENTRY( ptr, struct message, entry );
+        if (prev->msg != WM_POINTERUPDATE || !(prev_data = prev->data)) continue;
+        if (LOWORD(prev_data->rawinput.mouse.data) == LOWORD(msg_data->rawinput.mouse.data)) break;
+    }
+    if (!ptr) return 0;
+    if (prev->result) return 0;
+    if (prev->win && msg->win && prev->win != msg->win) return 0;
+    if (prev->type != msg->type) return 0;
+    /* now we can merge it */
+    prev->wparam  = msg->wparam;
+    prev->lparam  = msg->lparam;
+    prev->x       = msg->x;
+    prev->y       = msg->y;
+    prev->time    = msg->time;
+    prev_data->rawinput.mouse.data |= msg_data->rawinput.mouse.data;
+    prev_data->rawinput.mouse.x = msg_data->rawinput.mouse.x;
+    prev_data->rawinput.mouse.y = msg_data->rawinput.mouse.y;
+    list_remove( ptr );
+    list_add_tail( &input->msg_list, ptr );
+    return 1;
+}
+
 /* try to merge a WM_MOUSEMOVE message with the last in the list; return 1 if successful */
 static int merge_mousemove( struct thread_input *input, const struct message *msg )
 {
@@ -664,7 +694,7 @@ static int merge_mousemove( struct thread_input *input, const struct message *ms
     for (ptr = list_tail( &input->msg_list ); ptr; ptr = list_prev( &input->msg_list, ptr ))
     {
         prev = LIST_ENTRY( ptr, struct message, entry );
-        if (prev->msg != WM_INPUT) break;
+        if (prev->msg != WM_INPUT && prev->msg != WM_POINTERUPDATE) break;
     }
     if (!ptr) return 0;
     if (prev->result) return 0;
@@ -719,6 +749,7 @@ static int merge_message( struct thread_input *input, const struct message *msg
     if (msg->msg == WM_MOUSEMOVE) return merge_mousemove( input, msg );
     if (msg->msg == WM_WINE_CLIPCURSOR) return merge_unique_message( input, WM_WINE_CLIPCURSOR, msg );
     if (msg->msg == WM_WINE_SETCURSOR) return merge_unique_message( input, WM_WINE_SETCURSOR, msg );
+    if (msg->msg == WM_POINTERUPDATE) return merge_pointer_update_message( input, msg );
     return 0;
 }
 
@@ -2122,6 +2153,58 @@ static int queue_keyboard_message( struct desktop *desktop, user_handle_t win, c
     return wait;
 }
 
+struct touch
+{
+    struct list entry;
+    struct desktop *desktop;
+    user_handle_t win;
+    hw_input_t input;
+    struct timeout_user *timeout;
+};
+
+static void queue_touch_input_message( void *private )
+{
+    struct hw_msg_source source = { IMDT_UNAVAILABLE, IMDT_TOUCH };
+    struct touch *touch = private;
+    struct desktop *desktop = touch->desktop;
+    const hw_input_t *input = &touch->input;
+    user_handle_t win = touch->win;
+    struct hardware_msg_data *msg_data;
+    struct message *msg;
+
+    if (!(msg = alloc_hardware_message( 0, source, get_tick_count(), 0 ))) return;
+
+    msg_data = msg->data;
+    msg_data->info     = 0;
+    msg_data->size     = sizeof(*msg_data);
+    msg_data->flags    = input->hw.lparam;
+    msg_data->rawinput = input->hw.rawinput;
+
+    msg->win       = get_user_full_handle( win );
+    msg->msg       = input->hw.msg;
+    msg->wparam    = 0;
+    msg->lparam    = input->hw.lparam;
+    msg->x         = input->hw.rawinput.mouse.x;
+    msg->y         = input->hw.rawinput.mouse.y;
+
+    queue_hardware_message( desktop, msg, 1 );
+    touch->timeout = add_timeout_user( -160000, queue_touch_input_message, touch );
+}
+
+static struct touch *find_touch_input( struct desktop *desktop, unsigned int id )
+{
+    struct touch *touch;
+
+    LIST_FOR_EACH_ENTRY( touch, &desktop->touches, struct touch, entry )
+        if (LOWORD(touch->input.hw.rawinput.mouse.data) == id) return touch;
+
+    touch = mem_alloc( sizeof(struct touch) );
+    list_add_tail( &desktop->touches, &touch->entry );
+    touch->desktop = desktop;
+    touch->timeout = NULL;
+    return touch;
+}
+
 /* queue a hardware message for a custom type of event */
 static void queue_custom_hardware_message( struct desktop *desktop, user_handle_t win,
                                            unsigned int origin, const hw_input_t *input )
@@ -2129,6 +2212,7 @@ static void queue_custom_hardware_message( struct desktop *desktop, user_handle_
     struct hw_msg_source source = { IMDT_UNAVAILABLE, origin };
     struct hardware_msg_data *msg_data;
     struct rawinput_message raw_msg;
+    struct touch *touch;
     struct message *msg;
     data_size_t report_size = 0;
 
@@ -2178,6 +2262,21 @@ static void queue_custom_hardware_message( struct desktop *desktop, user_handle_
         msg_data->size     = sizeof(*msg_data);
         msg_data->flags    = input->hw.lparam;
         msg_data->rawinput = input->hw.rawinput;
+        touch = find_touch_input( desktop, LOWORD(input->hw.rawinput.mouse.data) );
+        if (touch->timeout) remove_timeout_user( touch->timeout );
+        if (input->hw.msg != WM_POINTERUP)
+        {
+            touch->win = win;
+            touch->input = *input;
+            touch->input.hw.msg = WM_POINTERUPDATE;
+            touch->input.hw.rawinput.mouse.data &= ~(POINTER_MESSAGE_FLAG_NEW << 16);
+            touch->timeout = add_timeout_user( -160000, queue_touch_input_message, touch );
+        }
+        else
+        {
+            list_remove( &touch->entry );
+            free( touch );
+        }
     }
 
     msg->win       = get_user_full_handle( win );
@@ -2504,6 +2603,21 @@ void post_win_event( struct thread *thread, unsigned int event,
     }
 }
 
+void free_touches( struct desktop *desktop, user_handle_t window )
+{
+    struct touch *touch, *next;
+
+    LIST_FOR_EACH_ENTRY_SAFE( touch, next, &desktop->touches, struct touch, entry )
+    {
+        if (!window || touch->win == window)
+        {
+            list_remove( &touch->entry );
+            if (touch->timeout) remove_timeout_user( touch->timeout );
+            free( touch );
+        }
+    }
+}
+
 /* free all hotkeys on a desktop, optionally filtering by window */
 void free_hotkeys( struct desktop *desktop, user_handle_t window )
 {
diff --git a/server/user.h b/server/user.h
index 8fa55e09b0f..f3936b7579c 100644
--- a/server/user.h
+++ b/server/user.h
@@ -72,6 +72,7 @@ struct desktop
     struct list          hotkeys;          /* list of registered hotkeys */
     struct timeout_user *close_timeout;    /* timeout before closing the desktop */
     timeout_t            close_timeout_val;/* timeout duration before closing desktop */
+    struct list          touches;          /* list of active touches */
     struct thread_input *foreground_input; /* thread input of foreground thread */
     unsigned int         users;            /* processes and threads using this desktop */
     user_handle_t                          cursor_win;       /* window that contains the cursor */
@@ -121,6 +122,7 @@ extern void post_win_event( struct thread *thread, unsigned int event,
                             const WCHAR *module, data_size_t module_size,
                             user_handle_t handle );
 extern void free_hotkeys( struct desktop *desktop, user_handle_t window );
+extern void free_touches( struct desktop *desktop, user_handle_t window );
 
 /* region functions */
 
diff --git a/server/window.c b/server/window.c
index a2b277d413a..bd11d0d32de 100644
--- a/server/window.c
+++ b/server/window.c
@@ -2026,6 +2026,7 @@ void free_window_handle( struct window *win )
     if (win == progman_window) progman_window = NULL;
     if (win == taskman_window) taskman_window = NULL;
     free_hotkeys( win->desktop, win->handle );
+    free_touches( win->desktop, win->handle );
     cleanup_clipboard_window( win->desktop, win->handle );
     destroy_properties( win );
     if (is_desktop_window(win))
diff --git a/server/winstation.c b/server/winstation.c
index 5903497d61e..33803ec1990 100644
--- a/server/winstation.c
+++ b/server/winstation.c
@@ -240,6 +240,7 @@ static struct desktop *create_desktop( const struct unicode_str *name, unsigned
             memset( desktop->keystate, 0, sizeof(desktop->keystate) );
             list_add_tail( &winstation->desktops, &desktop->entry );
             list_init( &desktop->hotkeys );
+            list_init( &desktop->touches );
             if (!init_desktop_mapping( desktop, name ))
             {
                 release_object( desktop );
@@ -292,6 +293,7 @@ static void desktop_destroy( struct object *obj )
     struct desktop *desktop = (struct desktop *)obj;
 
     free_hotkeys( desktop, 0 );
+    free_touches( desktop, 0 );
     if (desktop->top_window) free_window_handle( desktop->top_window );
     if (desktop->msg_window) free_window_handle( desktop->msg_window );
     if (desktop->global_hooks) release_object( desktop->global_hooks );
-- 
2.43.2

